### DISCUSSION TITLE
Node feature embedding are identical.

### DISCUSSION THREAD

**Discussion by lingchen1991** [pyg-team/pytorch_geometric#3747] [Q&A]
I used GATConv and SAGPooling to classify **a undirected complete graph with 48 nodes**. (torch-geometic version 1.6.1)

But I found that the output of **GATConv(512, 3)** shows that all 48 node feature embedding are the same.

tensor([[-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237],
        [-43.2148,  -0.8637,  -0.2237]], device='cuda:0',
       grad_fn=<AsStridedBackward>)

Then I looked into the source code of GATConv class, and found that the following line matters.

 **"        out = self.propagate(edge_index, x=(x_l, x_r), alpha=(alpha_l, alpha_r), size=size) "**

The inputs including **'x_l', 'x_r', 'alpha_l', 'alpha_r'** shows different values in terms of node. However, the **'out'** shows the identical 3D features in terms of node. Just like the above tensor.

Furthermore, I looked at the **self.propagate** function in the MessagePassing class. But the argument is no matched with the one used in GATConv class. 

I also looked into the value of **'alpha'** in the message function in GATConv class. The alpha are the totally the same with the value 1 / 48 = 0.0208.

![image](https://user-images.githubusercontent.com/36982036/147042534-bb4c8677-5f97-464d-bb13-7d3d6422444f.png)

Thus, I would like to ask you that do you have any clues on this situation?
And how the self.propagate function works? How come that the dim of **alpha_l/r** is 48 while the dim of **alpha_i/j** is 48*48?

BTW, in the original paper, there is a **weight vector a** for calculating the alphas by the two concatenated node embedding. Hence, the dim of **weight vector a** should be **two times of out_channels**. However, I cannot see any learnable parameter with this dim.
![image](https://user-images.githubusercontent.com/36982036/147044477-8d232a78-9dda-497d-8678-58426b7240b8.png)

The output attention weight alpha in above equation as shown below, which leads to the identical node feature embedding.
![image](https://user-images.githubusercontent.com/36982036/147060653-8ce5209c-716a-4b08-87bf-0e8ae231e684.png)

**Comment by rusty1s** [comment]
This is hard to tell without knowing your input feature matrix and `edge_index`. Obviously, this shouldn't be the case. IMO, this can only happen if the features for all nodes are the same (this also leads to uniform distribution of attention coefficients). Can you confirm? You can also try to upload the input feature matrix and `edge_index` for me to reproduce this issue.

PyG splits the weight vector `a` into two parts, such that we do not need to explicitly concatenate `Wh_i` and `Wh_j` (which saves some memory), *e.g.*, the following formulation is equivalent:
```
alpha = self.att_src @ Wh_i + self.att_dst @ Wh_j
```

**Reply by lingchen1991** [reply]
Thank you for your kindly response and the formulation.

 I confirm that the features for all nodes are different. I have uploaded part of my code and data for reproduction, which can reproduce in my machine. You can run by **"python test4.py --lr 1e-4 --dataset ./samples"** . Thanks again!

**Reply by rusty1s** [reply]
Thank you. Are attention coefficients also the same when actually training the model?

I can confirm that output embeddings are the same for an untrained GAT network. While the attention scores are different before the `softmax`, they indeed get squashed to `~1/48` after the softmax (which is, however, to be expected for an untrained neural network). One thing that may help is to increase the `negative_slope` of `GATConv` (e.g., to `0.5` or `1.0`).

**Reply by lingchen1991** [reply]
This is the alpha matrix [48, 48], which was saved after several iterations, of a sample.

![image](https://user-images.githubusercontent.com/36982036/147217011-3201aa0b-0568-4554-be18-f0dcaa4e0333.png)

This is the matrix for another sample, which saved after several epochs.

![image](https://user-images.githubusercontent.com/36982036/147219682-29a386b9-9074-4930-833b-442aaa3e5313.png)

The alpha matrix with all 0.208, which was shown in my original post, are the saved by output of SAGPooling. I also used GATConv to SAGPooling. Since the input of the SAGPooling are the graph whose node embedding feature are the same, all elements of the output alpha matrix are the same as 0.0208.

I also try to set the negative_slope to 1.0. But the issue still exists.

**Reply by rusty1s** [reply]
The main questions to me are: Does the model train? Does the loss decrease at all? If the model does not train, then it is to be expected that attention coefficients do not produce meaningful output. Can you shed some light on this? :)

**Reply by lingchen1991** [reply]
The model did train and the loss did decrease. But I cannot understand why the all 48 node embedding features have to converge to the same.

![image](https://user-images.githubusercontent.com/36982036/147242195-fb5256bc-39b3-4d3c-99af-c5399b7a8ef8.png)

**Reply by rusty1s** [reply]
Thanks for confirming. This is interesting. Even after 100 epochs, attention coefficients are nearly the same? What happens if you apply some standardization on the node input features to ensure that they are nicely distributed:
```python
data.x = data.x - data.x.mean(dim=0, keepdim=True) / data.x.std(dim=0, keepdim=True)
```
